sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-generation.service.ts

```diff
@@ -0,0 +1,506 @@
+import { Injectable } from "@nestjs/common";
+import { DISBURSEMENT_FILE_GENERATION_ANTICIPATION_DAYS } from "../../constants";
+import { addDays, round } from "@sims/utilities";
+import { EntityManager } from "typeorm";
+import {
+  AwardValueWithRelatedSchedule,
+  DisbursementOverawardService,
+  StudentRestrictionSharedService,
+} from "@sims/services";
+import {
+  ApplicationStatus,
+  COEStatus,
+  DisbursementSchedule,
+  OfferingIntensity,
+  RestrictionActionType,
+  mapFromRawAndEntities,
+  DisbursementScheduleStatus,
+  DisbursementValue,
+  DisbursementOveraward,
+  DisbursementValueType,
+  Student,
+  DisbursementOverawardOriginType,
+} from "@sims/sims-db";
+import { ECertDisbursementSchedule } from "./e-cert-generation.models";
+import { AwardOverawardBalance } from "@sims/services/disbursement-overaward/disbursement-overaward.models";
+import {
+  BC_FUNDING_TYPES,
+  BC_TOTAL_GRANT_AWARD_CODE,
+  LOAN_TYPES,
+} from "@sims/services/constants";
+import { SystemUsersService } from "@sims/services/system-users";
+
+/**
+ * While performing a possible huge amount of updates,
+ * breaks the execution in chunks.
+ */
+const DISBURSEMENT_SCHEDULES_UPDATE_CHUNK_SIZE = 1000;
+
+/**
+ * Manages all the preparation of the disbursements data needed to
+ * generate the e-Cert. Check and execute possible overawards deductions
+ * and calculate the awards effective values to be used to generate the e-Cert.
+ * All methods are prepared to be executed on a single transaction.
+ */
+@Injectable()
+export class ECertGenerationService {
+  constructor(
+    private readonly studentRestrictionService: StudentRestrictionSharedService,
+    private readonly disbursementOverawardService: DisbursementOverawardService,
+    private readonly systemUsersService: SystemUsersService,
+  ) {}
+
+  /**
+   * Get the information needed for the e-Cert generation, execute all
+   * calculations needed and return the records ready to generate the e-Cert file.
+   * @param offeringIntensity disbursement offering intensity.
+   * @param entityManager used to execute the commands in the same transaction.
+   * @returns records ready to generate the e-Cert file.
+   */
+  async prepareDisbursementsForECertGeneration(
+    offeringIntensity: OfferingIntensity,
+    entityManager: EntityManager,
+  ): Promise<ECertDisbursementSchedule[]> {
+    const disbursements = await this.getECertInformationToBeSent(
+      offeringIntensity,
+      entityManager,
+    );
+    if (!disbursements?.length) {
```

Comment:
`getECertInformationToBeSent` returns `mapFromRawAndEntities` so, it's null. when there are no records

Reply:
Are you saying that the code will not work or that you prefer `if (!disbursements)` only?

Reply:
2nd option... but if we have the plan to update `mapFromRawAndEntities ` then leave it.

Reply:
Yes, I believe that we should update the `mapFromRawAndEntities `. I can do it after this PR.